### DISCUSSION TITLE
Library/Module Drawer for Reusable Components

### DISCUSSION THREAD

**Discussion by bpartridge** [react-page/react-page#1201] [Ideas]
Would it be possible to implement a "library/module" system for reusable content?

https://stripo.email/blog/stripo-content-modules-what-are-they-and-why-do-you-need-them/ is an incredibly useful feature that the Stripo folks implemented (very similar spirit to this project, but specialized in email HTML rendering!), where you can create a hierarchy of components, hit Save As Module, and reuse it in future projects.

It's a much more powerful version of createInitialChildren; for instance, you could have HeroBasic and HeroWithCTA modules each of which have a HeroPlugin as their top level plugin.

Of course, everyone would have their own storage for this, and likely their own way of wanting to display and browse modules. But we could potentially add the right hooks into React Page by having:

- The ability to add custom tabs (as components passed to Editor as customizations) into the PluginDrawer with custom contents
- Documentation on how things you add there might want to call `useInsertNew -> insertNew(PartialCell)`
- An example of how one might override BottomToolbar to add a button that saves the current cell's data and descendants as a PartialCell, and since you're adding this component at the Editor level it can call a callback you give it at the Editor level.

Really, the only thing that doesn't exist here already is an extension point for the PluginDrawer... and even then one could implement their own AddModule drawer and link it to a customOptions button (which should be documented better).

And I'm not fully sure of the nuances of useInsertNew, since it doesn't seem to be intended to be used by non-internal code, but presumably it *could* be given a deep data structure that would take precedence over createInitialChildren? Since, I think, this is how duplication works?

Regardless, this seems like something that might be relevant to have some first-class support for in the core package, or at least could be an example!

**Comment by macrozone** [comment]
yes, thats exactly what i thought with https://github.com/react-page/react-page/issues/953

the api for that would probably be that you have to add additional props to the editor:

```
presets={...}
onSavePreset={...}

```

if `onSavePreset` is defined, it would automatically add a button to the bottom toolbar as you suggested, clicking that button would pass the current cell (or partial cell) onto onSavePreset, so developers can use that to save it somewhere and pass all available presets to `presets`

we would also need some additional callbacks like `onDeletePreset`.

presets would appear somewhere in the plugin drawer, which i think would need an overhowl anyway
